### DISCUSSION TITLE
How to use the body from the extractor Request<Body>

### DISCUSSION THREAD

**Discussion by R2-t** [tokio-rs/axum#1994] [Q&A] [answered]
Hey, I have been going through the axum docs, and they suggest to use `Request<Body>` whenever one wants to have more control over the request. But I have found it difficult to parse it into something else:

```rs
pub async fn handler(
    State(db_pool): State<AppState>,
    body: Request<Body>,
) -> Result<impl IntoResponse, ApiError> {
    println!("{:?}", body);
    let my_body_string = <Body as Into<String>>::into(*body.body());
    Ok(())
}
```

How can I make it usable? I don't find enough information about the type `Body`, and from `println!("{:?}")` I know it's a `Body(Streaming)`

**Comment by davidpdrsn** [comment]
You need to use the methods from the [`http_body::Body`](https://docs.rs/http-body/latest/http_body/trait.Body.html) trait. There is also [`hyper::body::to_bytes`](https://docs.rs/hyper/latest/hyper/body/fn.to_bytes.html) for buffering the whole body into a `Bytes` (don't forget to add a limit if you use this)

But from your code it seems like you just want a `String` in which case you can do this:

```rust
pub async fn handler(
    State(db_pool): State<AppState>,
    body: String,
) -> Result<impl IntoResponse, ApiError> {
    println!("{:?}", body);
    Ok(())
}
```

**Reply by R2-t** [reply]
If I decide to convert it to bytes, how can I manipulate those bytes later?

**Reply by davidpdrsn** [reply]
I don't understand what you mean. If you have the bytes then you just do stuff with them.

```rust
let bytes = string_body.into_bytes();
println!("there are {} bytes", bytes.len());
```

**Reply by R2-t** [reply]
I mean, if I want to convert the bytes to string, are they valid UTF-8?

**Reply by davidpdrsn** [reply]
All `String`s in rust are valid UTF-8 and it is undefined behavior (very very bad) otherwise.

**Reply by R2-t** [reply]
Hold on a second, Body can be converted to Byte, and then those bytes can be manipulated as a string?

**Reply by davidpdrsn** [reply]
No. Once you've converted something into bytes you need to call https://doc.rust-lang.org/stable/std/string/struct.String.html#method.from_utf8 to get the `String` back.

**Reply by R2-t** [reply]
> You need to use the methods from the [`http_body::Body`](https://docs.rs/http-body/latest/http_body/trait.Body.html) trait. There is also [`hyper::body::to_bytes`](https://docs.rs/hyper/latest/hyper/body/fn.to_bytes.html) for buffering the whole body into a `Bytes` (don't forget to add a limit if you use this)
> 
> But from your code it seems like you just want a `String` in which case you can do this:
> 
> ```rust
> pub async fn handler(
>     State(db_pool): State<AppState>,
>     body: String,
> ) -> Result<impl IntoResponse, ApiError> {
>     println!("{:?}", body);
>     Ok(())
> }
> ```

Well I had the chance recently to test this, and I don't see any to_bytes from `Request<Body>` it keeps complaining that there is not function for it...

**Reply by davidpdrsn** [reply]
Like I said it is [`hyper::body::to_bytes`](https://docs.rs/hyper/latest/hyper/body/fn.to_bytes.html).

You can also extract `axum::body::Bytes` if you need the raw bytes:

```rust
pub async fn handler(
    State(db_pool): State<AppState>,
    body: axum::body::Bytes,
) -> Result<impl IntoResponse, ApiError> {
    Ok(())
}
```

**Reply by R2-t** [reply]
Oh got it, that worked!

Now the question remains, from `Bytes` to `String`...

**Reply by R2-t** [reply]
Btw
```rs
pub async fn handler(
    State(db_pool): State<AppState>,
    body: Request<Body>,
) -> Result<impl IntoResponse, ApiError>{
    let my_body_bytes: String = body.into_body().into();
    Ok(())
}
```

Trying to cast to String from Body doesn't work

**Reply by davidpdrsn** [reply]
Sorry but I really don't understand what the problem is. If you want the body as a String then extract it as such. Otherwise extract Bytes.

**Reply by R2-t** [reply, marked as answer]
```rs
pub async fn handler(
    State(db_pool): State<AppState>,
    body: Request<Body>,
) -> Result<impl IntoResponse, ApiError> {
    let my_body_bytes = hyper::body::to_bytes(body.into_body())
        .await
        .map_err(|_| ApiError::InternalServerError)?;

    println!(
        "{}",
        String::from_utf8(my_body_bytes.to_vec()).map_err(|_| ApiError::InternalServerError)?
    );
    Ok(())
}
```

This is what I wanted... Yes I know that I can use String as an extractor, but I don't know how that magic happens, thus I wanted to implement it myself.

**Reply by davidpdrsn** [reply]
> Yes I know that I can use String as an extractor, but I don't know how that magic happens, thus I wanted to implement it myself.

Please say so next time.

**Reply by R2-t** [reply]
> > Yes I know that I can use String as an extractor, but I don't know how that magic happens, thus I wanted to implement it myself.
> 
> Please say so next time.

Sorry for that, I'll try to be more clear next time! Anyway, thank you for your time
